src/pkg/egress/syslog/https_batch.go

```diff
@@ -13,15 +13,120 @@ import (
 	"github.com/valyala/fasthttp"
 )
 
+// --- Coordinator definition ---
+type RetryCoordinator struct {
+	sem chan struct{}
+}
+
+var (
+	globalRetryCoordinator     *RetryCoordinator
+	globalRetryCoordinatorOnce sync.Once
+	maxParallelRetries         = 2
+)
+
+func WithParallelRetries(n int) {
+	maxParallelRetries = n
+	globalRetryCoordinatorOnce = sync.Once{}
+	globalRetryCoordinator = nil
+}
+
+func GetGlobalRetryCoordinator() *RetryCoordinator {
+	globalRetryCoordinatorOnce.Do(func() {
+		globalRetryCoordinator = &RetryCoordinator{
+			sem: make(chan struct{}, maxParallelRetries),
+		}
+	})
+	return globalRetryCoordinator
+}
+
+func (c *RetryCoordinator) Acquire() {
+	c.sem <- struct{}{}
+}
+
+func (c *RetryCoordinator) Release() {
+	<-c.sem
+}
+
+type InternalRetryWriter interface {
+	ConfigureRetry(retryDuration RetryDuration, maxRetries int)
+}
+
+type Retryer struct {
+	retryDuration RetryDuration
+	maxRetries    int
+	binding       *URLBinding
+	coordinator   *RetryCoordinator
+}
+
+func NewRetryer(
+	binding *URLBinding,
+	retryDuration RetryDuration,
+	maxRetries int,
+) *Retryer {
+	return &Retryer{
+		retryDuration: retryDuration,
+		maxRetries:    maxRetries,
+		binding:       binding,
+		coordinator:   GetGlobalRetryCoordinator(),
+	}
+}
+
+func (r *Retryer) Retry(batch []byte, msgCount float64, function func([]byte, float64) error) {
+	logTemplate := "failed to write to %s, retrying in %s, err: %s"
```

Comment:
IMHO, it is fine to have this log in two different places rather than using `logTemplate`.
This is also inconsistent with the "Context cancelled..." log which is printed in two places as well.

Reply:
will improve on this pattern, this is a remnant from where i got the logging logic from